File: DocumentAlterations/AICodingGenerator/ai-coding-context-generator.py
import os
import re
from typing import List, Dict

class AICodingContextGenerator:
    def __init__(self, project_root: str):
        self.project_root = project_root
        self.project_files = self._get_project_files()
        self.conversation_files = set()
        self.file_contents = {}

    def _get_project_files(self) -> List[str]:
        project_files = []
        for root, _, files in os.walk(self.project_root):
            for file in files:
                if file.endswith(('.py', '.js', '.html', '.css', '.json')):
                    project_files.append(os.path.join(root, file))
        return project_files

    def add_conversation_file(self, file_path: str):
        if os.path.exists(file_path):
            self.conversation_files.add(file_path)
        else:
            print(f"Warning: File {file_path} does not exist in the project.")

    def _read_file_content(self, file_path: str) -> str:
        if file_path not in self.file_contents:
            with open(file_path, 'r') as file:
                self.file_contents[file_path] = file.read()
        return self.file_contents[file_path]

    def _extract_functions(self, content: str) -> List[str]:
        function_pattern = r'def\s+(\w+)\s*\(.*?\):'
        return re.findall(function_pattern, content)

    def _extract_classes(self, content: str) -> List[str]:
        class_pattern = r'class\s+(\w+)(?:\(.*?\))?:'
        return re.findall(class_pattern, content)

    def generate_context(self) -> Dict[str, Dict]:
        context = {}

        for file_path in set(self.project_files) | self.conversation_files:
            file_content = self._read_file_content(file_path)
            functions = self._extract_functions(file_content)
            classes = self._extract_classes(file_content)

            context[file_path] = {
                'functions': functions,
                'classes': classes,
                'is_conversation_file': file_path in self.conversation_files
            }

        return context

    def provide_insights(self) -> str:
        context = self.generate_context()
        insights = []

        total_files = len(context)
        total_functions = sum(len(v['functions']) for v in context.values())
        total_classes = sum(len(v['classes']) for v in context.values())

        insights.append(f"Project Overview:")
        insights.append(f"- Total files analyzed: {total_files}")
        insights.append(f"- Total functions found: {total_functions}")
        insights.append(f"- Total classes found: {total_classes}")

        if self.conversation_files:
            insights.append("\nFiles mentioned in conversation:")
            for file in self.conversation_files:
                if file in context:
                    insights.append(f"- {file}")
                    insights.append(f"  Functions: {', '.join(context[file]['functions'])}")
                    insights.append(f"  Classes: {', '.join(context[file]['classes'])}")

        return "\n".join(insights)

# Example usage
if __name__ == "__main__":
    generator = AICodingContextGenerator("/path/to/project/root")
    generator.add_conversation_file("/path/to/project/root/specific_file.py")
    print(generator.provide_insights())

File: DocumentAlterations/AICodingGenerator/AICodingContextGenerator.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class AICodingContextGenerator {
    private final Path projectRoot;
    private final List<Path> projectFiles;
    private final Set<Path> conversationFiles;
    private final Map<Path, String> fileContents;

    public AICodingContextGenerator(String projectRoot) throws IOException {
        this.projectRoot = Paths.get(projectRoot);
        this.projectFiles = getProjectFiles();
        this.conversationFiles = new HashSet<>();
        this.fileContents = new HashMap<>();
    }

    private List<Path> getProjectFiles() throws IOException {
        return Files.walk(projectRoot)
                .filter(Files::isRegularFile)
                .filter(path -> {
                    String fileName = path.getFileName().toString();
                    return fileName.endsWith(".java") || fileName.endsWith(".js") ||
                            fileName.endsWith(".html") || fileName.endsWith(".css") ||
                            fileName.endsWith(".json");
                })
                .collect(Collectors.toList());
    }

    public void addConversationFile(String filePath) {
        Path path = Paths.get(filePath);
        if (Files.exists(path)) {
            conversationFiles.add(path);
        } else {
            System.out.println("Warning: File " + filePath + " does not exist in the project.");
        }
    }

    private String readFileContent(Path filePath) throws IOException {
        if (!fileContents.containsKey(filePath)) {
            fileContents.put(filePath, new String(Files.readAllBytes(filePath)));
        }
        return fileContents.get(filePath);
    }

    private List<String> extractFunctions(String content) {
        List<String> functions = new ArrayList<>();
        Pattern pattern = Pattern
                .compile("(?:public|private|protected|static|\\s) +[\\w\\<\\>\\[\\]]+\\s+(\\w+) *\\([^\\)]*\\) *\\{?");
        Matcher matcher = pattern.matcher(content);
        while (matcher.find()) {
            functions.add(matcher.group(1));
        }
        return functions;
    }

    private List<String> extractClasses(String content) {
        List<String> classes = new ArrayList<>();
        Pattern pattern = Pattern.compile("class\\s+(\\w+)");
        Matcher matcher = pattern.matcher(content);
        while (matcher.find()) {
            classes.add(matcher.group(1));
        }
        return classes;
    }

    public Map<Path, Map<String, Object>> generateContext() throws IOException {
        Map<Path, Map<String, Object>> context = new HashMap<>();

        Set<Path> allFiles = new HashSet<>(projectFiles);
        allFiles.addAll(conversationFiles);

        for (Path filePath : allFiles) {
            String fileContent = readFileContent(filePath);
            List<String> functions = extractFunctions(fileContent);
            List<String> classes = extractClasses(fileContent);

            Map<String, Object> fileContext = new HashMap<>();
            fileContext.put("functions", functions);
            fileContext.put("classes", classes);
            fileContext.put("isConversationFile", conversationFiles.contains(filePath));

            context.put(filePath, fileContext);
        }

        return context;
    }

    public String provideInsights() throws IOException {
        Map<Path, Map<String, Object>> context = generateContext();
        StringBuilder insights = new StringBuilder();

        int totalFiles = context.size();
        int totalFunctions = context.values().stream()
                .mapToInt(v -> ((List<?>) v.get("functions")).size())
                .sum();
        int totalClasses = context.values().stream()
                .mapToInt(v -> ((List<?>) v.get("classes")).size())
                .sum();

        insights.append("Project Overview:\n");
        insights.append("- Total files analyzed: ").append(totalFiles).append("\n");
        insights.append("- Total functions found: ").append(totalFunctions).append("\n");
        insights.append("- Total classes found: ").append(totalClasses).append("\n");

        if (!conversationFiles.isEmpty()) {
            insights.append("\nFiles mentioned in conversation:\n");
            for (Path file : conversationFiles) {
                if (context.containsKey(file)) {
                    Map<String, Object> fileContext = context.get(file);
                    insights.append("- ").append(file).append("\n");
                    insights.append("  Functions: ")
                            .append(String.join(", ", (List<String>) fileContext.get("functions"))).append("\n");
                    insights.append("  Classes: ").append(String.join(", ", (List<String>) fileContext.get("classes")))
                            .append("\n");
                }
            }
        }

        return insights.toString();
    }

    public static void main(String[] args) {
        try {
            AICodingContextGenerator generator = new AICodingContextGenerator("/path/to/project/root");
            generator.addConversationFile("/path/to/project/root/specific_file.java");
            System.out.println(generator.provideInsights());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
